File: app/src/main/java/com/mainstreetcode/teammate/fragments/registration/SignUpFragment.java
package com.mainstreetcode.teammate.fragments.registration;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.activities.RegistrationActivity;
import com.mainstreetcode.teammate.baseclasses.RegistrationActivityFragment;
import com.mainstreetcode.teammate.rest.TeammateService;
import com.mainstreetcode.teammate.util.ErrorHandler;
import com.tunjid.androidbootstrap.core.text.SpanBuilder;

import java.util.concurrent.TimeUnit;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.view.ViewCompat;
import io.reactivex.Completable;

import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;


public class SignUpFragment extends RegistrationActivityFragment
        implements
        TextView.OnEditorActionListener {

    private EditText firstNameInput;
    private EditText lastNameInput;
    private EditText emailInput;
    private EditText passwordInput;
    private CheckBox terms;

    public static SignUpFragment newInstance() {
        SignUpFragment fragment = new SignUpFragment();
        Bundle args = new Bundle();

        fragment.setArguments(args);
        fragment.setEnterExitTransitions();
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_sign_up, container, false);
        View border = rootView.findViewById(R.id.card_view_wrapper);
        firstNameInput = rootView.findViewById(R.id.first_name);
        lastNameInput = rootView.findViewById(R.id.last_name);
        emailInput = rootView.findViewById(R.id.email);
        passwordInput = rootView.findViewById(R.id.password);
        terms = rootView.findViewById(R.id.terms);

        passwordInput.setOnEditorActionListener(this);
        terms.setText(getTermsCharSequence());
        terms.setMovementMethod(LinkMovementMethod.getInstance());

        ViewCompat.setTransitionName(border, SplashFragment.TRANSITION_BACKGROUND);
        ViewCompat.setTransitionName(rootView.findViewById(R.id.member_info), SplashFragment.TRANSITION_TITLE);

        disposables.add(Completable.timer(200, TimeUnit.MILLISECONDS).observeOn(mainThread()).subscribe(() -> {
            rootView.findViewById(R.id.first_name_wrapper).setVisibility(View.VISIBLE);
            rootView.findViewById(R.id.last_name_wrapper).setVisibility(View.VISIBLE);
            rootView.findViewById(R.id.email_wrapper).setVisibility(View.VISIBLE);
            rootView.findViewById(R.id.password_wrapper).setVisibility(View.VISIBLE);
        }, ErrorHandler.EMPTY));

        return rootView;
    }

    @Override
    @StringRes
    protected int getFabStringResource() { return R.string.submit; }

    @Override
    @DrawableRes
    protected int getFabIconResource() { return R.drawable.ic_check_white_24dp; }

    @Override protected CharSequence getToolbarTitle() {
        return getString(R.string.sign_up);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        firstNameInput = null;
        lastNameInput = null;
        emailInput = null;
        passwordInput = null;
    }

    @Override
    public boolean showsFab() {
        return true;
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.fab:
                signUp();
                break;
        }
    }

    @Override
    public boolean onEditorAction(TextView textView, int actionId, KeyEvent event) {
        if ((actionId == EditorInfo.IME_ACTION_DONE) || ((event.getKeyCode() == KeyEvent.KEYCODE_ENTER)
                && (event.getAction() == KeyEvent.ACTION_DOWN))) {
            signUp();
            return true;
        }
        return false;
    }

    private void signUp() {
        boolean hasFirstName = VALIDATOR.isNotEmpty(firstNameInput);
        boolean hasLastName = VALIDATOR.isNotEmpty(lastNameInput);
        boolean hasEmail = VALIDATOR.isValidEmail(emailInput);
        boolean hasPassword = VALIDATOR.isValidPassword(passwordInput);
        boolean acceptedTerms = terms.isChecked();

        if (!acceptedTerms) showSnackbar(getString(R.string.sign_up_terms_accept));

        if (hasFirstName && hasLastName && hasEmail && hasPassword && acceptedTerms) {
            String firstName = firstNameInput.getText().toString();
            String lastName = lastNameInput.getText().toString();
            String email = emailInput.getText().toString();
            String password = passwordInput.getText().toString();

            toggleProgress(true);

            disposables.add(viewModel.signUp(firstName, lastName, email, password)
                    .subscribe(user -> onSignUp(), defaultErrorHandler)
            );
        }
    }

    private void onSignUp() {
        toggleProgress(false);
        hideKeyboard();
        RegistrationActivity.startMainActivity(getActivity());
    }

    private CharSequence getTermsCharSequence() {
        return SpanBuilder.format(getString(R.string.sign_up_terms_phrase),
                clickableSpan(getString(R.string.sign_up_terms), () -> startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(TeammateService.API_BASE_URL + "terms")))),
                clickableSpan(getString(R.string.sign_up_privacy_policy), () -> startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(TeammateService.API_BASE_URL + "privacy")))));
    }

    private CharSequence clickableSpan(CharSequence text, Runnable clickAction) {
        SpannableString spannableString = new SpannableString(text);
        ClickableSpan clickableSpan = new ClickableSpan() {
            public void onClick(View widget) {
                clickAction.run();
            }

            public void updateDrawState(TextPaint ds) {
                ds.setColor(Color.WHITE);
                ds.setUnderlineText(true);
            }
        };

        spannableString.setSpan(clickableSpan, 0, spannableString.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
        return spannableString;
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/fragments/registration/SignUpFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.fragments.registration

import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.text.SpannableString
import android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.baseclasses.TeammatesBaseFragment
import com.mainstreetcode.teammate.databinding.FragmentSignUpBinding
import com.mainstreetcode.teammate.rest.TeammateService
import com.mainstreetcode.teammate.util.ErrorHandler
import com.mainstreetcode.teammate.util.hasValidEmail
import com.mainstreetcode.teammate.util.hasValidName
import com.mainstreetcode.teammate.util.hasValidPassword
import com.mainstreetcode.teammate.util.input
import com.tunjid.androidx.core.text.SpanBuilder
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
import java.util.concurrent.TimeUnit


class SignUpFragment : TeammatesBaseFragment(R.layout.fragment_sign_up), TextView.OnEditorActionListener {

    private var binding: FragmentSignUpBinding? = null

    private val termsCharSequence: CharSequence
        get() = SpanBuilder.format(getString(R.string.sign_up_terms_phrase),
                clickableSpan(getString(R.string.sign_up_terms)) { startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(TeammateService.API_BASE_URL + "terms"))) },
                clickableSpan(getString(R.string.sign_up_privacy_policy)) { startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(TeammateService.API_BASE_URL + "privacy"))) })

    override val showsFab: Boolean = true

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) = FragmentSignUpBinding.bind(view).run {
        super.onViewCreated(view, savedInstanceState)
        binding = this

        defaultUi(
                fabText = R.string.submit,
                fabIcon = R.drawable.ic_check_white_24dp,
                fabShows = true,
                toolbarShows = false,
                bottomNavShows = false,
                grassShows = true,
                navBarColor = GRASS_COLOR
        )

        disposables.add(Completable.timer(200, TimeUnit.MILLISECONDS).observeOn(mainThread()).subscribe({
            firstNameWrapper.visibility = View.VISIBLE
            lastNameWrapper.visibility = View.VISIBLE
            emailWrapper.visibility = View.VISIBLE
            passwordWrapper.visibility = View.VISIBLE
        }, ErrorHandler.EMPTY::invoke))

        terms.text = termsCharSequence
        terms.movementMethod = LinkMovementMethod.getInstance()

        memberInfo.transitionName = SplashFragment.TRANSITION_TITLE
        cardViewWrapper.transitionName = SplashFragment.TRANSITION_BACKGROUND

        password.setOnEditorActionListener(this@SignUpFragment)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onClick(view: View) = when (view.id) {
        R.id.fab -> signUp()
        else -> Unit
    }

    override fun onEditorAction(textView: TextView, actionId: Int, event: KeyEvent): Boolean = when {
        actionId == EditorInfo.IME_ACTION_DONE || event.keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN -> signUp().let { true }
        else -> false
    }

    private fun signUp() {
        val binding = this.binding ?: return
        val hasFirstName = binding.firstName.hasValidName
        val hasLastName = binding.lastName.hasValidName
        val hasEmail = binding.email.hasValidEmail
        val hasPassword = binding.password.hasValidPassword
        val acceptedTerms = binding.terms.isChecked

        if (acceptedTerms) transientBarDriver.showSnackBar(getString(R.string.sign_up_terms_accept))

        if (hasFirstName && hasLastName && hasEmail && hasPassword && acceptedTerms) {
            val firstName = binding.firstName.input.toString()
            val lastName = binding.lastName.input.toString()
            val email = binding.email.input.toString()
            val password = binding.password.input.toString()

            transientBarDriver.toggleProgress(true)

            disposables.add(userViewModel.signUp(firstName, lastName, email, password)
                    .subscribe({ onSignUp() }, defaultErrorHandler::invoke)
            )
        }
    }

    private fun onSignUp() {
        transientBarDriver.toggleProgress(false)
        hideKeyboard()
        navigator.completeSignIn()
    }

    private fun clickableSpan(text: CharSequence, clickAction: () -> Unit): CharSequence {
        val spannableString = SpannableString(text)
        val clickableSpan = object : ClickableSpan() {
            override fun onClick(widget: View) = clickAction.invoke()

            override fun updateDrawState(ds: TextPaint) {
                ds.color = Color.WHITE
                ds.isUnderlineText = true
            }
        }

        spannableString.setSpan(clickableSpan, 0, spannableString.length, SPAN_EXCLUSIVE_EXCLUSIVE)
        return spannableString
    }

    companion object {

        fun newInstance(): SignUpFragment = SignUpFragment().apply {
            arguments = Bundle()
            setEnterExitTransitions()
        }
    }
}